sway/desktop/transaction.c

```diff
@@ -0,0 +1,395 @@
+			transaction->instructions->items[i];
+		if (instruction->container == container) {
+			return true;
+		}
+	}
+	return false;
+}
+
+void transaction_add_container(struct sway_transaction *transaction,
+		struct sway_container *container) {
+	if (transaction_has_container(transaction, container)) {
+		return;
+	}
+	struct sway_transaction_instruction *instruction =
+		calloc(1, sizeof(struct sway_transaction_instruction));
+	instruction->transaction = transaction;
+	instruction->container = container;
+
+	copy_pending_state(container, &instruction->state);
+
+	if (container->type == C_VIEW) {
+		save_view_buffer(container->sway_view, instruction);
+	}
+	list_add(transaction->instructions, instruction);
+}
+
+void transaction_add_damage(struct sway_transaction *transaction,
+		struct wlr_box *_box) {
+	struct wlr_box *box = calloc(1, sizeof(struct wlr_box));
+	memcpy(box, _box, sizeof(struct wlr_box));
+	list_add(transaction->damage, box);
+}
+
+/**
+ * Apply a transaction to the "current" state of the tree.
+ */
+static void transaction_apply(struct sway_transaction *transaction) {
+	wlr_log(L_DEBUG, "Applying transaction %p", transaction);
+	if (server.debug_txn_timings) {
+		struct timespec now;
+		clock_gettime(CLOCK_MONOTONIC, &now);
+		struct timespec *create = &transaction->create_time;
+		struct timespec *commit = &transaction->commit_time;
+		float ms_arranging = (commit->tv_sec - create->tv_sec) * 1000 +
+			(commit->tv_nsec - create->tv_nsec) / 1000000.0;
+		float ms_waiting = (now.tv_sec - commit->tv_sec) * 1000 +
+			(now.tv_nsec - commit->tv_nsec) / 1000000.0;
+		float ms_total = ms_arranging + ms_waiting;
+		wlr_log(L_DEBUG, "Transaction %p: %.1fms arranging, %.1fms waiting, "
+				"%.1fms total (%.1f frames if 60hz)", transaction,
+				ms_arranging, ms_waiting, ms_total, ms_total / (1000 / 60));
+	}
+	int i;
+	// Apply the instruction state to the container's current state
+	for (i = 0; i < transaction->instructions->length; ++i) {
+		struct sway_transaction_instruction *instruction =
+			transaction->instructions->items[i];
+		struct sway_container *container = instruction->container;
+
+		// There are separate children lists for each instruction state, the
+		// container's current state and the container's pending state
+		// (ie. con->children). The list itself needs to be freed here.
+		// Any child containers which are being deleted will be cleaned up in
+		// transaction_destroy().
+		list_free(container->current.children);
+
+		memcpy(&container->current, &instruction->state,
+				sizeof(struct sway_container_state));
+	}
+
+	// Apply damage
+	for (i = 0; i < transaction->damage->length; ++i) {
+		struct wlr_box *box = transaction->damage->items[i];
+		for (int j = 0; j < root_container.children->length; ++j) {
+			struct sway_container *output = root_container.children->items[j];
+			output_damage_box(output->sway_output, box);
+		}
+	}
+}
+
+static void progress_queue() {
```

Comment:
It's a problem we have a lot in sway but I think even static function names should have a semi-obvious name that says what they work on, `progress_queue` in particular might be difficult for someone to understand in a stack trace here. I'm as bad with names as anyone else though, `transaction_process` or something? (might turn out difficult to differentiate with apply?)

Well, just a thought.

Reply:
Yeah, I'm open to ideas. I don't think it should be something like `transaction_process` though because that implies it "processes a transaction" when in reality it iterates the queue and processes as many as it can. I think it should be `<something>_queue`.

Some ideas:

* `advance_queue`
* `check_queue`
* `handle_queue`
* `process_queue`
* `progress_queue`
* `run_queue`
* `work_queue`

You could also also add the word `transaction` to the above so it's `advance_transaction_queue` and so on.

Reply:
just `transaction_progress_queue` is fine then, it's the main thing I find could be confusing.